### PR TITLE
Feat/merge new tree builder

### PR DESCRIPTION
**Pull request by ChickenLover**
## Describe the changes

This PR adds CUDA code for the poseidon and the tree builder

### CODE REVIEW

**Review by DmytroTym — APPROVED** [review]
One data point: our depth `2^30` tree currently takes about 14.4 seconds on a 3090Ti machine versus 13.6 for the supranational's implementation

File: icicle/appUtils/poseidon/optimized/poseidon.cu [hunk 1]
```diff
@@ -0,0 +1,163 @@
...
+  {
+    shared_states[threadIdx.x] = element;
+    __syncthreads();
+
+    typename S::Wide element_wide = S::mul_wide(shared_states[vec_number * size], matrix[element_number]);
+    for (int i = 1; i < size; i++) {
+      element_wide =
+        element_wide + S::mul_wide(shared_states[vec_number * size + i], matrix[i * size + element_number]);
+    }
+    __syncthreads();
```
[the remark this answers is not in the document]
Comment (ChickenLover) [reply icicle/appUtils/poseidon/optimized/poseidon.cu:51]: it doesn't work without this magic sync :) We can file an issue for that

File: icicle/appUtils/poseidon/optimized/poseidon.cu [hunk 2]
```diff
@@ -0,0 +1,163 @@
...
+    S* sparse_matrix = &config.sparse_matrices[(config.t * 2 - 1) * round_number];
+
+    typename S::Wide state_0_wide = S::mul_wide(element, sparse_matrix[0]);
+    for (int i = 1; i < config.t; i++) {
+      state_0_wide = state_0_wide + S::mul_wide(state[i], sparse_matrix[i]);
+    }
+    state[0] = S::reduce(state_0_wide);
+
+    for (int i = 1; i < config.t; i++) {
+      state[i] = state[i] + (element * sparse_matrix[config.t + i - 1]);
```
[the remark this answers is not in the document]
Comment (ChickenLover) [reply icicle/appUtils/poseidon/optimized/poseidon.cu:123]: Good idea, moved to #308

**Review by jeremyfelder — COMMENTED** [review]
Change base branch to `main`. We don't use `dev` anymore

**Comment by ChickenLover** [comment]
> Change base branch to `main`. We don't use `dev` anymore

We actually do now. This PR contains some bugs so it will only be merged after I deal with them

**Pull request closed without merging**